src/profiler.cpp

```diff
@@ -1647,6 +1652,124 @@ void Profiler::dumpText(Writer& out, Arguments& args) {
     }
 }
 
+/*
+ * If the value has been seen already, return its idx in `values`.
+ * If not, update the map and return the index.
+ */
+static inline u32 getOrSetIndex(
+        std::unordered_map<std::string, u32>& idx_map,
+        const std::string& value
+) {
+    return idx_map.insert({value, idx_map.size()}).first->second;
+}
+
+void Profiler::dumpOtlp(Writer& out, Arguments& args) {
+    // _otlp_buffer contains a ProfileData message
+    _otlp_buffer.reset();
+
+    std::unordered_map<std::string, u32> string_idx_map;
+    std::unordered_map<std::string, u32> function_idx_map;
+
+    protobuf_mark_t resource_profiles_mark = _otlp_buffer.startMessage(Otlp::ProfilesData::resource_profiles);
+    protobuf_mark_t scope_profiles_mark = _otlp_buffer.startMessage(Otlp::ResourceProfiles::scope_profiles);
+    protobuf_mark_t profile_mark = _otlp_buffer.startMessage(Otlp::ScopeProfiles::profiles);
+
+    _otlp_buffer.field(Otlp::Profile::period, (u64) args._interval);
+
+    protobuf_mark_t sample_type_mark = _otlp_buffer.startMessage(Otlp::Profile::sample_type);
+    _otlp_buffer.field(Otlp::ValueType::type_strindex, getOrSetIndex(string_idx_map, _engine->type()));
+    _otlp_buffer.field(Otlp::ValueType::unit_strindex, getOrSetIndex(string_idx_map, _engine->units()));
+    _otlp_buffer.field(Otlp::ValueType::aggregation_temporality, Otlp::AggregationTemporality::cumulative);
+    _otlp_buffer.commitMessage(sample_type_mark);
+
+    std::vector<CallTraceSample*> samples;
+    _call_trace_storage.collectSamples(samples);
+
+    u32 samples_counter_cache[samples.size()];
+    u32 samples_num_frames_cache[samples.size()];
```

Comment:
It's not "cache", this is just a temporary holder. `samples_counter` and `samples_num_frames` will be good enough.